File: blender_roblox_sync/server/format_types/str.py
import struct

format = struct.Struct("<H")
format_size = format.size
unpack_from = format.unpack_from
pack = format.pack

def read(self, format_data, **masks):
	str_len, = unpack_from(*self.read_buffer(format_size))
	str_value, = struct.unpack_from(f"<{str_len}s", *self.read_buffer(str_len))
	return (str_value.decode("utf-8"),)

def write(self, args, args_count, format_data, **masks):
	str_value = args[args_count]
	str_len = len(str_value)
	self.write_buffer(pack(str_len), format_size)
	self.write_buffer(struct.pack(f"<{str_len}s", str_value.encode("utf-8")), str_len)
	return args_count + 1

File: roblox_plugin/src/Server/formatTypes/str.ts
//!native
import type * as ProcessFormats from "../processFormats";
import SendMessagesThread from "../sendMessages";
import ReceiveMessagesThread from "../receiveMessages";

export function read(
	receiveThread: ReceiveMessagesThread,
	args: defined[],
	formatData: ProcessFormats.FormatData,
	masks: Map<string, boolean>,
) {
	const [readBuffer, readOffset] = ReceiveMessagesThread.readBuffer(receiveThread, 2);
	const strLen = buffer.readu16(readBuffer, readOffset);
	const [readBuffer2, readOffset2] = ReceiveMessagesThread.readBuffer(receiveThread, strLen);
	args.push(buffer.readstring(readBuffer2, readOffset2, strLen));
}
export function write(
	sendThread: SendMessagesThread,
	args: defined[],
	argsCount: number,
	formatData: ProcessFormats.FormatData,
	masks: Map<string, boolean>,
): number {
	const value = args[argsCount] as string;
	const strLen = value.size();

	const data = buffer.create(2 + strLen);
	buffer.writeu16(data, 0, strLen);
	buffer.writestring(data, 2, value, strLen);
	SendMessagesThread.writeBuffer(sendThread, data, 2 + strLen);
	return argsCount + 1;
}
